oauthproxy.go

```diff
@@ -571,22 +571,37 @@ func (p *OAuthProxy) ManualSignIn(req *http.Request) (string, bool) {
 func (p *OAuthProxy) GetRedirect(req *http.Request) (redirect string, err error) {
 	err = req.ParseForm()
 	if err != nil {
-		return
+		return redirect, err
 	}
 
 	redirect = req.Header.Get("X-Auth-Request-Redirect")
 	if req.Form.Get("rd") != "" {
 		redirect = req.Form.Get("rd")
 	}
 	if !p.IsValidRedirect(redirect) {
+		proto := ""
+		host := ""
 		// Use RequestURI to preserve ?query
-		redirect = req.URL.RequestURI()
-		if strings.HasPrefix(redirect, p.ProxyPrefix) {
-			redirect = "/"
+		uri := req.URL.RequestURI()
+
+		// Quirk: On reverse proxies that doesn't have support for
+		// "X-Auth-Request-Redirect" header or dynamic header/query string
+		// manipulation (like Traefik v1 and v2), we can try if the header
+		// X-Forwarded-Host exists or not.
+		if p.IsForwardedRequest(req) {
+			proto = fmt.Sprintf("%s://", util.GetRequestProto(req))
+			host = util.GetRequestHost(req)
+			uri = util.GetRequestURI(req)
+		}
+
+		if strings.HasPrefix(uri, p.ProxyPrefix) {
+			uri = "/"
 		}
+
+		redirect = fmt.Sprintf("%s%s%s", proto, host, uri)
```

Comment:
Maybe needs a test for non-standard ports? I believe the header for this isn't currently handled.

Checking here, Traefik seems to add this as `X-Forwarded-Port` rather than include it in the host.
https://github.com/traefik/traefik/blob/a3327c4430cf4a2cb35216776c2c9843281dec00/pkg/middlewares/auth/forward.go#L223-L225

Reply:
Hm, reading over the part I linked again, it seems actually Traefik prefers to include the port in the host. But if there is a trusted proxy before traefik, it passes the port header along. Either way, there may be cases where the port is split into a separate header.

Reply:
I believe `X-Forwarded-Host` should reflect the original `Host` header which would include an optional port if a non-canonical port was used, right?

I think we should rely solely on `X-Forwarded-Host` and not inject in `X-Forwarded-Port` here because there are a lot of areas that consume `X-Forwarded-Host` (cookies, logs, redirect). I think they should all be aligned and consistent.

What is the scenario you are thinking where an `X-Forwarded-Port` that deviates from `X-Forwarded-Host` would arise in viable architectures?

Reply:
I've looked around a bit and yes, Traefik behaves as you describe, port included in `X-Forwarded-Host` when it's the reverse proxy that connects to the client. But it would copy a `X-Forwarded-Port` header if a trusted proxy were in front of Traefik that sets it. There wasn't an example server that I could find mentioned in Traefik's commit history for this.

Since we don't know what server would use `X-Forwarded-Port` we can keep the existing assumption it should be using `X-Forwarded-Host` instead.

Reply:
Alright! I'll be committing to reflect that.